documentation for making a SC quest (#3789)
streetcomplete/StreetComplete: CONTRIBUTING_A_NEW_QUEST.md

- clone your fork of a StreetComplete repository
- open StreetComplete in Android Studio
- [setup an emulator in Android Studio](https://developer.android.com/studio/run/emulator#install) (you can also [connect to a real device via USB](https://developer.android.com/studio/run/device), this can replace using the emulator)
- run StreetComplete in the emulator (to verify that everything was setup as required)

If you are doing it for the first time, don't worry if there is an error to solve along the way, this is typical for setting up an Android development environment. See [CONTRIBUTING file](CONTRIBUTING.md#development) which has some links to information about the setup.

# Invent a new quest

## Own ideas
To [repeat](CONTRIBUTING.md#developing-new-quests) from that documentation file:  [**open an issue** discussing the quest](CONTRIBUTING.md#developing-new-quests), before starting other work. This way it can be confirmed that such quest can be included. This can be skipped if you are an [experienced](https://github.com/streetcomplete/StreetComplete/discussions/3450) StreetComplete contributor.

## Existing proposals

> You can also look at [quest proposals waiting for implementation](https://github.com/streetcomplete/StreetComplete/issues?q=is%3Aissue+is%3Aopen+label%3A%22new+quest%22).

Reviewer: Both as a general question and a comment on the doc: Does "quest proposals waiting for implementation" mean that every issue that has the label "New Quest" is meant to be viewed as "approved as okay and just waiting to be implemented"? (obviously excluding issues where additionally "Blocked" is applied) 
On that note, maybe the link could be extended to exclude issues with that label.
```suggestion
You can also look at [quest proposals waiting for implementation](https://github.com/streetcomplete/StreetComplete/issues?q=is%3Aissue+is%3Aopen+label%3A%22new+quest%22+-label%3A%22blocked%22).
```

Author: > Does "quest proposals waiting for implementation" mean that every issue that has the label "New Quest" is meant to be viewed as "approved as okay and just waiting to be implemented"?

yes - is it necessary to make it more clear?

> blocked

well, blocked ones can progress by submitting work upstream.

Not sure is it necessary to exclude them. Some of it is easier than actually making SC quest

Reviewer: Ah ok, I always thought that the new quest label simply means "This issue proposes a new quest" rather than "This issue proposes a new quest that was already determined to be a good fit and can be implemented"
> well, blocked ones can progress by submitting work upstream.
Oh well, that shows another misunderstanding on my side of the meaning of the issue lables

Maybe (if you deem this to be a misunderstanding more people could experience) a section could be added explaining what the different lables mean.

Reviewer 2: I added descriptions to the labels themselves as a first measure: https://github.com/streetcomplete/StreetComplete/labels

Author: I opened https://github.com/streetcomplete/StreetComplete/discussions/3840

Some quests are entirely defined in a single file, some have additional answer class, custom interface or utility classes.

# Adjust the copy

After the quest is copied it is necessary to adjust it a bit.

Change its class name and the file name to the new one.

In copied code change package info (things like `package de.westnordost.streetcomplete.quests.defibrillator` at the top) to match the new folder containing the quest.

# Add the quest to the list of active ones

Adjust [QuestsModule.kt](app/src/main/java/de/westnordost/streetcomplete/quests/QuestsModule.kt) file. It contains a big list of active quests, ordered by priority. Read [what governs their priority](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/QuestsModule.kt#L169-L192) but do not worry too much, it can be tweaked later.

Add your quest to the list so that it will be loaded by the app.

## Quest anatomy
See for example [simple yes/no quest asking whether AED is indoor or outdoor](app/src/main/java/de/westnordost/streetcomplete/quests/defibrillator/AddIsDefibrillatorIndoor.kt).

> TODO: maybe linking [app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmElementQuestType.kt](app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmElementQuestType.kt) would be better? And extend documentation there?

Author: Flagging TODO

### Element selection
elementFilter property defines nodes, ways and relations which will be selected for a given quest. It is an element selection used by OsmFilterQuestType.

```
"""
        nodes with
         emergency = defibrillator
         and access !~ private|no
         and !indoor
"""
``` 

This query will be limited to nodes (`nodes with`), which fulfill some requirements.

- `emergency = defibrillator` tag must be present
- `access` tag must not have values `private` or `no` to skip ones where mapper will be unable to survey (`!~ private|no` - to be more specific `optionA|optionB` is treated like `^optionA|optionB$` [regexp](https://en.wikipedia.org/wiki/Regular_expression))
- `indoor` key must not be present at all, to show only ones where this tag is still missing

It is specified as a string, in syntax specific to StreetComplete. You can look around some quests to see how it works. If you are trying to implement a new quest and you got stuck here, [open a new issue](https://github.com/streetcomplete/StreetComplete/issues) to request more thorough documentation here.

> 

Reviewer 3: It should be emphasized that in `elementFilter` syntax (unlike vast majority of other programming languages), there is **no precendence of `and` over `or`**,  so users should use a lot of parenthesis, or it will lead to puzzling bugs like https://github.com/streetcomplete/StreetComplete/pull/3784#issuecomment-1053674700

Reviewer 3: You mean in the parsing code? Sure, either of those would be ideal fix by me! But if that is too complicated to implement, it should at least be documented here in the docs.

Author: > You mean in the parsing code?

Yes, I think that SC crashing with error on start would be preferable and avoid troubles you had

Author: I opened https://github.com/streetcomplete/StreetComplete/issues/3838

Reviewer 4: This is wrong. There is operator precedence.

* Is quest listed?
  * Look at the quest list in settings - is your quest appearing there? If not - see [this step](#add-the-quest-to-the-list-of-active-ones).
  * While you are there you can disable all quests except yours for easier testing.
* Is it shown for expected elements?
  * Note that due to emulator peculiarities you may need to move map after quests are downloaded to see their markers ([gory details for curious](https://github.com/streetcomplete/StreetComplete/issues/2780))
* Is it selected for some unwanted elements?
* Can you tap on a quest marker to open quest form?
* Can you fill the answer as expected?
* Can you solve the quest?
* Is expected tagging being applied?
  * You can look at logs for info what was applied or use undo menu from bottom-left in the app itself.
  * You can freely answer - as long as you are not logged in, nothing will be submitted. Even after logging in you can disable uploading answers in settings.
* Is quest disappearing after being solved?
  * If no - then either tags are not being applied or element selection filter have some problems.

See "logcat" (bottom left area of the screen) to see stacktrace or logging messages.

## Adding logs

> `Log.w("Unique string for easy grepping in logcat", "Message with whatever you need like #${someVariable.itsProperty}")`

Reviewer 3: ```suggestion
`import android.util.Log`
`Log.w("Unique string for easy grepping in logcat", "Message with whatever you need like #${someVariable.itsProperty}")`
```

```
    override fun getTitle(tags: Map<String, String>) =
        if (hasFeatureName(tags))
            R.string.quest_wheelchairAccess_name_type_title
        else
            R.string.quest_wheelchairAccess_name_title

    override fun getTitleArgs(tags: Map<String, String>, featureName: Lazy<String?>): Array<String> {
        val name = tags["name"] ?: tags["brand"]
        return arrayOfNotNull(name, featureName.value)
    }
``` 

- `getTitleArgs` tries to pass to the title
  - name of the object
  - brand of the object (only if `name` tag is not used, and `brand` tag is present)
  - feature name (such as "greengrocer", translated to the proper language, based on [iD presets](https://github.com/westnordost/osmfeatures))

[Filtering in the element selection](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/wheelchair_access/AddWheelchairAccessBusiness.kt#L20) ensures that every object will have either `name` or `brand` tag.

> Though description from iD presets is not guaranted.

Reviewer 5: ```suggestion
Though description from iD presets is not guaranteed.
```

- familiarity with StreetComplete as a user - it is highly recommended to be familiar with how StreetComplete works. Making hundreds of edits is not mandatory, but having a passing familiarity with how surveying with StreetComplete works is needed.
- time - Android Studio setup is sadly complicated
  - Multiple hours are typical, though mostly spent on waiting for various downloads
  - setting up Android Studio will likely be more complex than writing the code
- [basic ability to use git](https://git-scm.com/docs/user-manual) (but as mentioned, anything GitHub specific will be explained)
  - there is version control integration available in Android Studio
    - From the Android Studio **VCS** menu, click **Enable Version Control** Integration ([see official docs](https://developer.android.com/studio/intro#version_control_basics))
    - Go to File -> Settings -> Version control in Android Studio to optionally configure git integration
  - Using git from command line or terminal in Android Studio also works fine
- GitHub [account](https://github.com/signup)
- around 15GB of free space on disk

>   - the initial setup will typically download half of the Internet while getting emulators and dependencies

Author: attempted joke (as in "downloading 50% of the Internet")

>   - the initial setup will typically download half of the Internet while getting emulators and dependencies

Author: Maybe worth removing as confusing but I like it, so will keep for now

>   - the initial setup will typically download half of the Internet while getting emulators and dependencies

Reviewer 2: Let's make it explicit then:

```suggestion
  - the initial setup will typically download half the Internet while getting emulators and dependencies
```

Add your quest to the list so that it will be loaded by the app.

As this point you can run the app in emulator - everything should work and one of quests will appear twice.

See [this step](https://github.com/matkoniecz/StreetComplete_quest_creation_tutorial/commit/b4f01eeb752f92fd6927fbb9a943ea19a4799eec) in the example repository.

# Modify the template

At this point prepared template can be modified to achieve the intended effect.

See for example [simple yes/no quest asking whether AED is indoor or outdoor](app/src/main/java/de/westnordost/streetcomplete/quests/defibrillator/AddIsDefibrillatorIndoor.kt).

## Element selection

`elementFilter` property defines nodes, ways and relations which will be selected for a given quest. It is an element selection used by OsmFilterQuestType.

```kotlin
    override val elementFilter = """
        nodes with
         emergency = defibrillator
         and access !~ private|no
         and !indoor
    """
```

This query will be limited to object which fulfill some requirements.

- `nodes with`
  - nodes only, ways and relations are not eligible
- `emergency = defibrillator`
  - this tag must be present
- `and access !~ private|no`

>   - and `access` tag must not have values `private` or `no`
>   - `!~ private|no` is transformed into `^optionA|optionB$` [regexp](https://en.wikipedia.org/wiki/Regular_expression)

Reviewer 2: ```suggestion
  - and `access` tag must not have values `private` or `no` (the `~` indicates it is a [regular expression](https://en.wikipedia.org/wiki/Regular_expression) which will be matched against the whole string)
```

>   - `!~ private|no` is transformed into `^optionA|optionB$` [regexp](https://en.wikipedia.org/wiki/Regular_expression)

Reviewer 4: better yet

>   - `!~ private|no` is transformed into `^optionA|optionB$` [regexp](https://en.wikipedia.org/wiki/Regular_expression)

Author: From looking into it:

`.ame ~ y.s` will not match `name ~ yess`

similarly 

`name ~ yes|no` will not match `yess`

This is different from for example Python regexp which will happily match on part of string and needs extra parts of regexp or additional checks to match only on entire text. While Kotlin has [matches](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/-regex/matches.html) and a separate [matchEntire](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/-regex/match-entire.html)

- Don't rely on the "leave a note" fallback too much. It is not intended and does not work as a regular answer but is designed to cover the case that the question was invalid itself because it was based on wrong data like i.e. the place does not exist anymore.
- The information the user should fill in should be as atomic as possible. Users are impatient, i.e. do not let them fill out a whole address with street name etc. when just the house number is fine too.
- "A picture is worth a thousand words": Often the term for certain things may not be enough to convey the meaning of certain predefined answers. Do you know what a wheelbender is? You will know if you see the photo.

Considerations about the edge cases to consider, how the design could look like and finding good representative photos or icons that match in style is also part of the preparational work that can be done without programming knowledge.

## New photos

Some quests will require photos for their interface.

> Photos must be available on open license - so not every photo found on the Internet is usable. You can either take your own photo or find an existing freely licensed photo.

Author: What about say https://wiki.openstreetmap.org/wiki/Template:WTFPL and other similar

though strictly speaking it is not claiming that no other license is acceptable...

> Photos must be available on open license - so not every photo found on the Internet is usable. You can either take your own photo or find an existing freely licensed photo.

Reviewer 4: You can append "and other similar licenses if you wish

```kotlin
override val elementFilter = """
    nodes, ways with amenity = motorcycle_parking
      and access !~ private|no
      and (!capacity or capacity older today -4 years)
"""
```

This quest will be triggered when:

- `nodes, ways with`
  - on nodes or ways (relations not eligible)
- `amenity = motorcycle_parking`
  - where `amenity = motorcycle_parking` tag is present
- `and access !~ private|no`
  - and `access` tag does not have value `private` nor `no`
- `and (!capacity or capacity older today -4 years)`
  - and one of following is fullfilled:
    - `capacity` tag is not present at all (`!capacity`)
    - element was not edited for a long time (base time is 4 years, but it can be influenced by user changing settings)
    - `check_date:capacity` with a date indicating that it is outdated (the same as above applies)

## Custom filters

It is possible to use far more complex filters when querying for eligible elements.

Matches like `surface ~ earth|dirt|ground` are possible and are evaluated as "`surface` is either of `earth`, `dirt`, `ground`"

`access !~ private|no` will be evaluated to "`access` is neither `private` nor `no`"

> Simple lists are [evaluated as sets](https://github.com/streetcomplete/StreetComplete/blob/2e812b9a3b5288983309a7edde6e8f9db05ad3f2/app/src/test/java/de/westnordost/streetcomplete/data/elementfilter/filters/ElementFilterOverpassKtTest.kt#L79-L88) - and for example `surface ~ earth|dirt|ground` will not match `surface=earther` tag. Basically, `surface ~ earth|dirt|ground` is treated as `surface ~ ^earth|dirt|ground$`

Reviewer 4: ```suggestion
```

Delete this because as noted before, the information is not relevant and also misleading

Good places to find freely licensed images are [Geograph](https://www.geograph.org.uk/) and [Wikimedia Commons](https://commons.wikimedia.org/).

You can also take your own photos, a standard smartphone camera is good enough. And sometimes a highly specific image is needed.

In StreetComplete many images have unusual composition. Often it is necessary to leave space for a label at the bottom.

Images should be free of visual debris, not misleading. Though it is fine to use an image not strictly matching what is depicted, as long as it is clear. For example, [a permanent pile of soil blocking road](https://wiki.openstreetmap.org/wiki/Tag:barrier%3Ddebris) is illustrated by a [temporary landslide](https://commons.wikimedia.org/wiki/File:Landslide_on_OR_42S_(46849629014).jpg). This is OK as images are illustrative.

Photos go to a different folder than SVGs: they can be used directly by the build process so put them into folders

- [mdpi](app/src/main/res/drawable-mdpi) - 384 pixels for images, with three square images in each row it would be 128 x 128 pixels for each)

> - [hdpi](app/src/main/res/drawable-hdpi) - 576 pixels for images (192 x 192 pixels in case of three quare images in each row)
> - [xhdpi](app/src/main/res/drawable-xhdpi) 768 pixels for images (256 x 256 pixels in case of three quare images in each row)
> - [xxhdpi](app/src/main/res/drawable-xxhdpi) 1152 pixels (384 x 384 pixels in case of three quare images in each row)

Reviewer 3: ```suggestion
- [hdpi](app/src/main/res/drawable-hdpi) - 576 pixels for images (192 x 192 pixels in case of three square images in each row)
- [xhdpi](app/src/main/res/drawable-xhdpi) 768 pixels for images (256 x 256 pixels in case of three square images in each row)
- [xxhdpi](app/src/main/res/drawable-xxhdpi) 1152 pixels (384 x 384 pixels in case of three square images in each row)
```

typos

Reviewer 2: This typo was reintroduced in the last commit.

  - Setup of environment takes a long time
  - Editing even the simplest quest requires edits to at least three different files
  - Kotlin familiarity is not required, but the ability to adjust to a new syntax is needed. If you are unfamiliar with Kotlin reading about [null safety related syntax](https://kotlinlang.org/docs/null-safety.html) is likely to be very useful.
  - It is expected that someone trying this can search for a solution to a typical problem on encountering an error message
- familiarity with StreetComplete as a user - it is highly recommended to be familiar with how StreetComplete works. Making hundreds of edits is not mandatory, but having a passing familiarity with how surveying with StreetComplete works is needed.
- time - Android Studio setup is sadly complicated
  - Multiple hours are typical, though mostly spent on waiting for various downloads
  - setting up Android Studio will likely be more complex than writing the code
- [basic ability to use git](https://git-scm.com/docs/user-manual) (but as mentioned, anything GitHub specific will be explained)
  - there is version control integration available in Android Studio

>     - From the Android Studio **VCS** menu, click **Enable Version Control Integration** ([see official docs](https://developer.android.com/studio/intro#version_control_basics))

Author: Not sure, but official documentation claims that you need to enable it.

Maybe I will setup AS in VM?

>     - From the Android Studio **VCS** menu, click **Enable Version Control Integration** ([see official docs](https://developer.android.com/studio/intro#version_control_basics))

Reviewer 4: Or maybe drop that whole section. Knowing how to use Git is a basic requirement.

    - Go to File -> Settings -> Version control in Android Studio to optionally configure git integration
  - Using git from command line or terminal in Android Studio also works fine
- GitHub [account](https://github.com/signup)
- around 15GB of free space on disk
  - the initial setup will typically download half the Internet while getting emulators and dependencies

# Dependencies - initial setup

- install necessary software ([Android Studio](https://developer.android.com/studio) and [git](https://git-scm.com/downloads))

> - create a GitHub [account](https://github.com/signup) if needed

Author: > the installer will certainly check if that space is available before installing.

large part of that goes to emulator. On initial setup I managed to run into space issues (not vert large SSD) and what worse emulator failed with cryptic errors that failed to explain anything.

> - create a GitHub [account](https://github.com/signup) if needed

Reviewer 4: That happend with your old ancient laptop from ??? you still had in 2018 though, right?

> - create a GitHub [account](https://github.com/signup) if needed

Author: Now, on my recent one (happened after for unrelated reasons I had 20 VMs setup which managed to eat nearly entire space)

> - create a GitHub [account](https://github.com/signup) if needed

Reviewer 4: Is this relevant? Who is going to setup 20 VMs, especially in the first-time setup?

```kotlin
override val elementFilter = """
    nodes, ways with amenity = motorcycle_parking
      and access !~ private|no
      and (!capacity or capacity older today -4 years)
"""
```

This quest will be triggered when:

- `nodes, ways with`
  - on nodes or ways (relations not eligible)
- `amenity = motorcycle_parking`
  - where `amenity = motorcycle_parking` tag is present
- `and access !~ private|no`
  - and `access` tag does not have value `private` nor `no`
- `and (!capacity or capacity older today -4 years)`
  - and one of following is fullfilled:
    - `capacity` tag is not present at all (`!capacity`)
    - element was not edited for a long time (base time is 4 years, but it can be influenced by user changing settings)
    - `check_date:capacity` with a date indicating that it is outdated (the same as above applies)

## Custom filters

It is possible to use far more complex filters when querying for eligible elements.

Matches like `surface ~ earth|dirt|ground` are possible and are evaluated as "`surface` is either of `earth`, `dirt`, `ground`"

`access !~ private|no` will be evaluated to "`access` is neither `private` nor `no`"

> Simple lists are [evaluated as sets](https://github.com/streetcomplete/StreetComplete/blob/2e812b9a3b5288983309a7edde6e8f9db05ad3f2/app/src/test/java/de/westnordost/streetcomplete/data/elementfilter/filters/ElementFilterOverpassKtTest.kt#L79-L88) - and for example `surface ~ earth|dirt|ground` will not match `surface=earther` tag. Basically, `surface ~ earth|dirt|ground` is treated as `surface ~ ^earth|dirt|ground$`

Reviewer 4: This is not really correct. It is simply a "matches entire string", not a "find in". That if the regex only contains "|" as regex character, the expression is optimized to run faster is an implementation detail.
In any case, the link is not helpful

Author: Which part is not correct here? Because I am really unsure what is supposed to be incorrect.

Reviewer 4: > Simple lists are evaluated as sets

is an implementation and the link that is supposed to prove that claim(?) links to code that is unrelated to that claim.

> `surface ~ earth|dirt|ground` is treated as `surface ~ ^earth|dirt|ground$`

is not really correct. `surface ~ earth|dirt|ground` is treated as `surface ~ earth|dirt|ground`. As written, **all** regexes used here are RegEx.matches, not RegEx.find. I.e. entire string needs to match the regex. Maybe you want to say "is treated as ... if it was written in OverpassQL Wizard"?

> But using regexp like `surface ~ ^(.*)[0-9]$` is [also possible](https://github.com/streetcomplete/StreetComplete/blob/master/app/src/test/java/de/westnordost/streetcomplete/data/elementfilter/filters/ElementFilterOverpassKtTest.kt#L79-L88).

Reviewer 4: I doubt it. The link is not helpful and misleading.

Reviewer 4: ```suggestion
```
So this should be deleted.

TODO: maybe linking [app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmElementQuestType.kt](app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmElementQuestType.kt) would be better? And extend documentation there?

### Element selection

elementFilter property defines nodes, ways and relations which will be selected for a given quest. It is an element selection used by OsmFilterQuestType.

```kotlin
"""
    nodes with
     emergency = defibrillator
     and access !~ private|no
     and !indoor
"""
```

This query will be limited to nodes (`nodes with`), which fulfill some requirements.

- `emergency = defibrillator` tag must be present
- `access` tag must not have values `private` or `no` to skip ones where mapper will be unable to survey (`!~ private|no` - to be more specific `optionA|optionB` is treated like `^optionA|optionB$` [regexp](https://en.wikipedia.org/wiki/Regular_expression))
- `indoor` key must not be present at all, to show only ones where this tag is still missing

> It is specified as a string, in syntax specific to StreetComplete. You can look around some quests to see how it works. If you are trying to implement a new quest and you got stuck here, [open a new issue](https://github.com/streetcomplete/StreetComplete/issues) to request more thorough documentation here.

Reviewer 3: It should be mentioned that `!` (not) operator is not supported in combination with parenthesis, see https://github.com/streetcomplete/StreetComplete/pull/3865#discussion_r825499349

Reviewer 3: undefined tag handling should also be explained; see https://github.com/streetcomplete/StreetComplete/pull/3865#discussion_r825499349

Reviewer 4: https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/data/elementfilter/ElementFilterExpression.kt#L25-L61

###### Resurvey

Some quests are asked not only when tag is missing but also when it is likely to be outdated.

Typical code is in [quest asking about motorcycle parking capacity](app/src/main/java/de/westnordost/streetcomplete/quests/motorcycle_parking_capacity/AddMotorcycleParkingCapacity.kt):

```kotlin
override val elementFilter = """
    nodes, ways with amenity = motorcycle_parking
      and access !~ private|no
      and (!capacity or capacity older today -4 years)
"""
```

This quest will be triggered when:

- on nodes and ways
- where `amenity = motorcycle_parking` tag is present
- `access` tag does not have value `private` nor `no`
- and one of following is fullfilled:
  - `capacity` tag is not present at all (`!capacity`)

>   - element was not edited for a long time (base time is 4 years, but it can by influenced by user changing settings)

Reviewer 5: ```suggestion
  - element was not edited for a long time (base time is 4 years, but it can be influenced by user changing settings)
```

After full release it will reach the entire StreetComplete audience who now will be able to more easily contribute to OpenStreetMap. Thanks in advance for that!

# Bad documentation is a bug

Unclear documents, including this one, are a bug. Feel free to either submit a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) with a fix or [open an issue](https://github.com/streetcomplete/StreetComplete/issues/new) describing your confusion.

Note that not everything will be directly described. This document very intentionally doesn't include a step-by-step guide to installing Android Studio, [linking](CONTRIBUTING.md#development) to official docs instead.

# More complexity

What was described above is an attempt to cover all aspects of quest creation, without describing all the complexity.

Below is some additional info.

> ## getTitle and getTitleArgs

Author: this just got simplified
